src/appleseed-max-impl/appleseedvolumemtl/appleseedvolumemtl.cpp

```diff
@@ -0,0 +1,593 @@
+        ParamBlockRefVolumeMtl,                        // parameter block's reference number
+
+        // --- P_AUTO_UI arguments for Volume rollup ---
+        IDD_FORMVIEW_VOLUME_PARAMS,                    // ID of the dialog template
+        IDS_FORMVIEW_VOLUME_PARAMS_TITLE,              // ID of the dialog's title string
+        0,                                          // IParamMap2 creation/deletion flag mask
+        0,                                          // rollup creation flag
+        nullptr,                                    // user dialog procedure
+        
+        // --- Parameters specifications for Volume rollup ---
+
+        ParamIdVolumeAbsorption, L"volume_absorption", TYPE_RGBA, P_ANIMATABLE, IDS_VOLUME_ABSORTION,
+            p_default, Color(0.5f, 0.5f, 0.5f),
+            p_ui, TYPE_COLORSWATCH, IDC_VOLUME_ABSORTION,
+        p_end,
+
+        ParamIdVolumeAbsorptionMultiplier, L"volume_absorption_multiplier", TYPE_FLOAT, P_ANIMATABLE, IDS_VOLUME_ABSORTION_MULTIPLIER,
+            p_default, 100.0f,
+            p_range, 0.0f, 100.0f,
+            p_ui, TYPE_SLIDER, EDITTYPE_FLOAT, IDC_EDIT_VOLUME_ABSORTION_MULTIPLIER, IDC_SLIDER_VOLUME_ABSORTION_MULTIPLIER, 10.0f,
+        p_end,
+
+        ParamIdVolumeScattering, L"volume_scattering", TYPE_RGBA, P_ANIMATABLE, IDS_VOLUME_SCATTERING,
+            p_default, Color(0.5f, 0.5f, 0.5f),
+            p_ui, TYPE_COLORSWATCH, IDC_VOLUME_SCATTERING,
+        p_end,
+
+        ParamIdVolumeScatteringMultiplier, L"volume_scattering_multiplier", TYPE_FLOAT, P_ANIMATABLE, IDS_VOLUME_SCATTERING_MULTIPLIER,
+            p_default, 100.0f,
+            p_range, 0.0f, 100.0f,
+            p_ui, TYPE_SLIDER, EDITTYPE_FLOAT, IDC_EDIT_VOLUME_SCATTERING_MULTIPLIER, IDC_SPINNER_VOLUME_SCATTERING_MULTIPLIER, SPIN_AUTOSCALE,
+        p_end,
+
+        ParamIdVolumeAverageCosine, L"volume_average_cosine", TYPE_FLOAT, P_ANIMATABLE, IDS_VOLUME_AVERAGE_COSINE,
+            p_default, 0.0f,
+            p_range, -1.0f, 1.0f,
+            p_ui, TYPE_SLIDER, EDITTYPE_FLOAT, IDC_EDIT_VOLUME_AVERAGE_COSINE, IDC_SPINNER_VOLUME_AVERAGE_COSINE, SPIN_AUTOSCALE,
+        p_end,
+
+        ParamIdVolumePhaseFunctionModel, L"volume_phase_function_model", TYPE_INT, 0, IDS_VOLUME_PHASE_FUNCTION_MODEL,
+            p_ui, TYPE_INT_COMBOBOX, IDC_COMBO_VOLUME_PHASE_FUNCTION_MODEL,
+            2, IDS_VOLUME_HENYEY_MODEL, IDS_VOLUME_ISOTROPIC_MODEL,
+            p_vals, 0, 1,
+            p_default, 0,
+        p_end,
+
+        // --- The end ---
+        p_end);
+}
+
+Class_ID AppleseedVolumeMtl::get_class_id()
+{
+    return Class_ID(0x5c2e11d7, 0x34c07107);
+}
+
+AppleseedVolumeMtl::AppleseedVolumeMtl()
+  : m_pblock(nullptr)
+{
+    m_params_validity.SetEmpty();
+
+    g_appleseed_volumemtl_classdesc.MakeAutoParamBlocks(this);
+}
+
+BaseInterface* AppleseedVolumeMtl::GetInterface(Interface_ID id)
+{
+    return
+        id == IAppleseedMtl::interface_id()
+        ? static_cast<IAppleseedMtl*>(this)
```

Comment:
Indent once more:

    return
        id == IAppleseedMtl::interface_id()
            ? static_cast<IAppleseedMtl*>(this)
            : Mtl::GetInterface(id);